File: site/js/mods/lfo4.js
/**
 * A fourth LFO: a fourth page under `[MOD]` that behaves like LFO1-3.
 *
 * The browser half of `src/dnfw/mods/lfo4.py`, which carries the evidence and
 * the hardware result. Same steps, same order: check the stock length and every
 * edit's stock bytes, rebuild the relocated parameter table from this image,
 * put it in the appended blob, write the edits, append the blob.
 * `scripts/js_lfo4_check.mjs` fails if the bytes differ from the Python's, and
 * the Python's are checked byte for byte against the release build when
 * `scripts/gen_lfo4_code.py` packages it.
 *
 * **No stock table bytes ship.** The relocated table is this image's own 320
 * records followed by LFO4's ten, which are this image's LFO3 records with five
 * fields rewritten -- the port of `dnfw.patch.paramtable.records` and
 * `dnfw.patch.lfo4records.build`. So a mod applied earlier that edits the stock
 * table (fxmod, moddest) is carried into the copy, **provided it is applied
 * first**: this one goes last.
 */

import { CODE } from "./lfo4-code.js";

export const ID = "lfo4";
export const NAME = "A fourth LFO";
export const SECTION = 3;
const BASE = 0x40000400;

// dnfw.patch.paramtable
const TABLE = 0x401f7fc8, RECORD = 60, COUNT = 320;
// dnfw.patch.lfo4records
const LFO3_FIRST = 94, GROUP_SIZE = 10, SLOT0 = 101;
const SLOT = 12, NRPN = 36, UNIQUE = 40, FLAGS = 44, PAGE_NAME = 52;
const UNSET = 0xffffffff;
const DEST_POSITION = 3, DEST_FLAGS = 0x00008000;
const SLOTS = [0, 1, 2, 3, 4, 5, 5, 6, 7, 1];

export class ModError extends Error {}

const hex = (s) => Uint8Array.from(s.match(/../g) ?? [], (b) => parseInt(b, 16));
const toHex = (a) => Array.from(a, (b) => b.toString(16).padStart(2, "0")).join("");
/** Python's `f"{n:,}"`, whatever the browser's locale. */
const grouped = (n) => String(n).replace(/\B(?=(\d{3})+(?!\d))/g, ",");
const u32 = (d, at) => ((d[at] << 24) >>> 0) + (d[at + 1] << 16) + (d[at + 2] << 8) + d[at + 3];
const put32 = (d, at, v) => {
  d[at] = v >>> 24; d[at + 1] = (v >>> 16) & 0xff; d[at + 2] = (v >>> 8) & 0xff; d[at + 3] = v & 0xff;
};

export const EDITS = CODE.edits.length;
export const EDITED_BYTES = CODE.edits.reduce((n, e) => n + e.new.length / 2, 0);
export const APPENDED = CODE.blob.length / 2;
export const RECORDS = CODE.table_records;

export function extents() {
  return [
    ...CODE.edits.map((e) => ({ section: SECTION, start: e.va - BASE,
                                length: e.new.length / 2, what: "LFO4 edit" })),
    { section: SECTION, start: CODE.area_va - BASE, length: APPENDED,
      what: "appended data area (loader, LFO4's C, the relocated table)" },
  ];
}

/** `want` values of field +40 inside the range the table uses and claimed by none. */
export function freeUniqueIds(content, want = GROUP_SIZE) {
  const used = new Set();
  for (let i = 0; i < COUNT; i++) used.add(u32(content, TABLE - BASE + RECORD * i + UNIQUE));
  used.delete(UNSET);
  let lo = Infinity, hi = -Infinity;
  for (const v of used) { if (v < lo) lo = v; if (v > hi) hi = v; }
  const gaps = [];
  for (let v = lo; v < hi; v++) if (!used.has(v)) gaps.push(v);
  if (gaps.length < want) {
    throw new ModError(`only ${gaps.length} unused id(s) inside the table's range, ${want} wanted`);
  }
  return gaps.slice(0, want);
}

/** LFO4's ten records: LFO3's, with the slot, NRPN, id, flags and page name rewritten. */
export function lfo4Records(content, pageNameVa) {
  const ids = freeUniqueIds(content);
  const out = new Uint8Array(RECORD * GROUP_SIZE);
  for (let k = 0; k < GROUP_SIZE; k++) {
    const at = TABLE - BASE + RECORD * (LFO3_FIRST + k);
    const record = content.slice(at, at + RECORD);
    put32(record, SLOT, SLOT0 + SLOTS[k]);
    put32(record, NRPN, UNSET);
    put32(record, UNIQUE, ids[k]);
    put32(record, FLAGS, k === DEST_POSITION ? DEST_FLAGS : 0);
    put32(record, PAGE_NAME, pageNameVa);
    out.set(record, RECORD * k);
  }
  return out;
}

/** The relocated table, from this image: its 320 records and LFO4's ten. */
export function table(content) {
  const nameVa = CODE.table_va + RECORD * CODE.table_records;
  const stock = content.subarray(TABLE - BASE, TABLE - BASE + RECORD * COUNT);
  const out = new Uint8Array(RECORD * (COUNT + GROUP_SIZE));
  out.set(stock);
  out.set(lfo4Records(content, nameVa), stock.length);
  return out;
}

/** Throw unless `content` is stock 1.11 MAIN OS wherever this mod writes. */
export function guard(content) {
  if (content.length !== CODE.stock_length) {
    throw new ModError(`MAIN OS is ${grouped(content.length)} B, not ${grouped(CODE.stock_length)}: `
      + "either not Digitone II 1.11, or another mod has already appended data "
      + "(lfowaves and bootscreen do)");
  }
  for (const e of CODE.edits) {
    const at = e.va - BASE;
    const have = toHex(content.subarray(at, at + e.stock.length / 2));
    if (have !== e.stock) {
      throw new ModError(`0x${e.va.toString(16).padStart(8, "0")} is not stock (${have.slice(0, 24)}...); `
        + "another mod has changed it, or this is not Digitone II 1.11");
    }
  }
}

/** Stock-guarded edits, then the appended area with the table filled in. */
export function compose(content) {
  guard(content);
  const blob = hex(CODE.blob);
  const records = table(content);
  const at = CODE.table_offset;
  if (blob.subarray(at, at + records.length).some((b) => b !== 0)) {
    throw new ModError("lfo4_code.json is damaged: the table's place is not blank");
  }
  blob.set(records, at);

  const out = new Uint8Array(content.length + blob.length);
  out.set(content);
  for (const e of CODE.edits) out.set(hex(e.new), e.va - BASE);
  out.set(blob, content.length);
  return out;
}

/** Throw unless this image is one the mod applies to; -> MAIN OS. */
export function check(firmware) {
  const section = firmware.container.find(SECTION);
  if (section === null) throw new ModError("image has no MAIN OS section");
  const original = section.unpack();
  if (original === null) throw new ModError("MAIN OS did not depack");
  guard(original);
  return original;
}

/** -> `{ content, notes }`, the new section 3. */
export function apply(firmware) {
  const original = check(firmware);
  const content = compose(original);
  return {
    content,
    notes: [`${EDITS} edits, appended ${grouped(APPENDED)} B; the parameter table rebuilt `
            + `from this image; MAIN OS ${grouped(content.length)} B`],
  };
}

File: src/dnfw/mods/lfo4.py
"""A fourth LFO: a fourth page under `[MOD]` that behaves like LFO1-3.

**Passed on the instrument** (`docs/lfo4-build-plan.md`): the page opens with
the stock LFO controls, its `DEST` browser works, it modulates on every voice,
and its settings are kept by an explicit SAVE PROJECT and survive a power-cycle
without one.

## What it is

C, in `csrc/lfo4/`, compiled and linked by `scripts/build_lfo4_browser.py`
into a chunk the start-up loader copies to RAM, plus the in-image edits that
reach it: the parameter table moved to make room for ten records, the page
list, the four hook sites into the C, and the engine stubs. The C remains the
source. `scripts/gen_lfo4_code.py` runs that build and records what it changed
(`lfo4_code.json`), and checks that this mod applied to stock reproduces the
build **byte for byte**, so what applies here is exactly what was gated.

## What does not ship

The relocated parameter table is 320 stock records followed by LFO4's ten,
which are LFO3's with five fields rewritten. None of it is in the JSON: it is
rebuilt at apply time from the image being modified, through the same library
functions the build uses. A mod applied earlier that edits the stock table
(`moddest` opens thirteen masks there) is therefore carried into the
relocated copy -- **provided it is applied first**, which is why the CLI
applies this mod last.

## What it shares

It uses the start-up hook and the appended data area, like `lfowaves` and
`bootscreen`, and requires MAIN OS to end where stock 1.11 ends. So it
cannot be combined with either of them yet: `docs/mods-compatibility.md`.
"""

from __future__ import annotations

import json
import pathlib

from . import Extent, ModError, Result
from ..patch import lfo4records, paramtable

ID = "lfo4"
NAME = "A fourth LFO"
SUMMARY = ("A fourth LFO page under [MOD], like LFO1-3: modulates on every voice, "
           "kept by a save and across a power-cycle.")
DEVICE = 0x15                      # Digitone II
SECTION = 3                        # MAIN OS
BASE = 0x40000400
APPLY_LAST = True                  # it copies the stock table: see the docstring

# What this mod copies out of the image, so a mod that edits it must come
# first: its edit would otherwise land on a table nothing reads any more.
COPIES = [Extent(SECTION, paramtable.TABLE - BASE, paramtable.RECORD * paramtable.COUNT,
                 "the stock parameter table, copied into the appended area")]

_CODE = pathlib.Path(__file__).with_name("lfo4_code.json")
SPEC = json.loads(_CODE.read_text()) if _CODE.exists() else None


def _spec() -> dict:
    if SPEC is None:
        raise ModError("lfo4_code.json is missing: run scripts/gen_lfo4_code.py")
    return SPEC


def extents(firmware=None) -> list[Extent]:
    spec = _spec()
    out = [Extent(SECTION, e["va"] - BASE, len(e["new"]) // 2, "LFO4 edit") for e in spec["edits"]]
    out.append(Extent(SECTION, spec["area_va"] - BASE, len(spec["blob"]) // 2,
                      "appended data area (loader, LFO4's C, the relocated table)"))
    return out


def _table(content: bytes) -> bytes:
    """The relocated table, from this image: its 320 records and LFO4's ten."""
    spec = _spec()
    name_va = spec["table_va"] + paramtable.RECORD * spec["table_records"]
    return (paramtable.records(content, BASE)
            + lfo4records.build(content, BASE, page_name_va=name_va))


def compose(content: bytes) -> bytes:
    """Stock-guarded edits, then the appended area with the table filled in."""
    spec = _spec()
    if len(content) != spec["stock_length"]:
        raise ModError(f"MAIN OS is {len(content):,} B, not {spec['stock_length']:,}: either not "
                       "Digitone II 1.11, or another mod has already appended data "
                       "(lfowaves and bootscreen do)")
    for e in spec["edits"]:
        at = e["va"] - BASE
        have = bytes(content[at:at + len(e["stock"]) // 2]).hex()
        if have != e["stock"]:
            raise ModError(f"0x{e['va']:08x} is not stock ({have[:24]}...); another mod "
                           "has changed it, or this is not Digitone II 1.11")

    blob = bytearray.fromhex(spec["blob"])
    records = _table(content)
    at = spec["table_offset"]
    if any(blob[at:at + len(records)]):
        raise ModError("lfo4_code.json is damaged: the table's place is not blank")
    blob[at:at + len(records)] = records

    out = bytearray(content)
    for e in spec["edits"]:
        at = e["va"] - BASE
        new = bytes.fromhex(e["new"])
        out[at:at + len(new)] = new
    return bytes(out + blob)


def apply(firmware) -> Result:
    section = firmware.container.find(SECTION)
    if section is None:
        raise ModError("image has no MAIN OS section")
    content = compose(section.unpack())
    spec = _spec()
    notes = [f"{len(spec['edits'])} edits, appended {len(spec['blob']) // 2:,} B; "
             f"the parameter table rebuilt from this image; MAIN OS {len(content):,} B"]
    return Result({SECTION: content}, extents(firmware), notes)
